### ISSUE TITLE
Improvements for big CI setups

### ISSUE THREAD

**Issue by PragTob** [simplecov-ruby/simplecov#964]
With #960 we got a big blocker for big CI setups (hundreds of parallel CI jobs) out of the way to use `SimpleCov.collate` and generate a report.

We still have plenty of optimization potential left to make SimpleCov faster and more efficient for those setups:

* [ ] allow saving of `.resultset.json` without going through `SimpleCov::Result` _or_ reducing the overhead of `SimpleCov::Result` massively, #960 already massively reduced how often we instantiate it but as its core we only want to dump the collected coverage data into `.resultset.json` which we should be able to do easily without it
* [ ] make it easy to run with no formatter, right now we recommend to run with a simple formatter but it'd be even better to run with no formatter at all - all the separate CI jobs don't need to go through the overhead of generating a full report, only the final one with `collate` should. I'm not even sure if we can specify `formatter nil` or `formatters []` right now.
* [ ] Package up these settings in a new profile for easy usage

**Comment by mhenrixon** [comment]
Hi @PragTob, I am interested in learning more about what I can do in the meanwhile. I upgraded to 0.21.2 because I saw performance was improved. Alas, the performance is considerably less for now. I am running the test suite locally (not in parallel) and by the SimpleCov starts to process the coverage I have an additional couple of minutes for SimpleCov to sort and format the result I guess.

I turned off branch coverage, which helped a little but the performance is still far from acceptable to me. Do you need any assistance in finding and fixing these issues? If so, can you point me to where I can start looking? I saw #916 for instance, might be a good place to start?

**Comment by PragTob** [comment]
@mhenrixon can you please open a new issue about this? With your setup and at best a repo to reproduce/where the time is spent. The changes _should_ not have impacted normal setups that don't use `SimpleCov.collate` as the changes were aimed at making those faster. Of course, I might have broken stuff while changing this which of course I'd wanna change. Performance should not have digressed from 0.20 to 0.21. Or do you mean even different scenarios? Anyhow, a separate issue would be appreciated - thank you!

**Comment by sferik** [comment]
Two of the three bullets are now done:

- **No-formatter mode** landed in https://github.com/simplecov-ruby/simplecov/pull/1188. `formatter false` (or `formatters []`) opts a worker out of formatting entirely and `Result#format!` returns `nil`. Workers can drop their `.resultset.json` and exit without paying for HTML or multi-formatter overhead. A final `SimpleCov.collate` step formats once.
- **Reducing `Result` overhead** was addressed earlier by https://github.com/simplecov-ruby/simplecov/commit/fc7db4dc24706d3ebc2108f85d49d5846f20e269: `Result.new` is roughly 7× faster on already-string-keyed input (the `collate` hot path), because it no longer round-trips coverage data through `JSON.parse(JSON.dump(...))` (see https://github.com/simplecov-ruby/simplecov/issues/916).

I didn’t implement the third bullet (**packaging these settings as a profile**), since it’s just a one-line configuration.

Closing this issue, as the substantive optimizations are in, but feel free to open a new issue or PR if you still think the profile is worth doing.

This will ship in the v1.0.0 release but you can test it today by replacing the `simplecov` line in your `Gemfile` with:
 
```ruby
gem "simplecov", git: "https://github.com/simplecov-ruby/simplecov.git", branch: "main", require: false
```

I’d greatly appreciate if you’re able to test this before the v1.0.0 release and report and issues you find.
